File: tests/d_polyglot/programs/pure/zig/collatz.zig
// Pure WASM: Compute Collatz sequence length for n=27 = 112 steps
// Tests: loops, conditionals, division

fn collatz_length(start: i32) i32 {
    var n: i64 = start; // Use i64 to avoid overflow
    var steps: i32 = 0;

    while (n != 1) {
        if (@mod(n, 2) == 0) {
            n = @divTrunc(n, 2);
        } else {
            n = 3 * n + 1;
        }
        steps += 1;
    }
    return steps;
}

export fn main() i32 {
    // n=27 has 111 steps (longest for n < 100)
    // Return modulo 256 to fit in exit code
    return @mod(collatz_length(27), 256); // 111
}

File: tests/d_polyglot/programs/pure/c/collatz.c
// Pure WASM: Compute Collatz sequence length for n=27 = 111 steps
// Tests: loops, conditionals, division

static int collatz_length(int start) {
    long long n = start; // Use 64-bit to avoid overflow
    int steps = 0;

    while (n != 1) {
        if (n % 2 == 0) {
            n /= 2;
        } else {
            n = 3 * n + 1;
        }
        steps++;
    }
    return steps;
}

__attribute__((export_name("main")))
int main(void) {
    // n=27 has 111 steps (longest for n < 100)
    return collatz_length(27) % 256; // 111
}
